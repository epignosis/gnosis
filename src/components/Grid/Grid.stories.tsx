import { Story } from "@storybook/react";
import React from "react";
import Grid from "./Grid";

export default {
  component: Grid,
  title: "Components/Grid",
};

export const Simple: Story = () => (
  <Grid templateColumns={[1, 2, 3, 4, 5]} gap={2}>
    <div style={{ backgroundColor: "whitesmoke", height: "100%" }}>Item 1</div>
    <div style={{ backgroundColor: "whitesmoke", height: "100%" }}>Item 2</div>
    <div style={{ backgroundColor: "whitesmoke", height: "100%" }}>Item 3</div>
    <div style={{ backgroundColor: "whitesmoke", height: "100%" }}>Item 4</div>
    <div style={{ backgroundColor: "whitesmoke", height: "100%" }}>Item 5</div>
    <div style={{ backgroundColor: "whitesmoke", height: "100%" }}>Item 6</div>
    <div style={{ backgroundColor: "whitesmoke", height: "100%" }}>Item 7</div>
    <div style={{ backgroundColor: "whitesmoke", height: "100%" }}>Item 8</div>
  </Grid>
);

Simple.parameters = {
  controls: { hideNoControlsWarning: true },
};

export const WithCustomGridItems: Story = () => (
  <Grid templateColumns={[1, 2, 3, 4, 5]} gap={2}>
    <Grid.Item colSpan={[1, 2, 1]}>
      <div style={{ backgroundColor: "whitesmoke", height: "100%" }}>Item 1</div>
    </Grid.Item>
    <Grid.Item colSpan={[1, 2, null, null, 1]}>
      <div style={{ backgroundColor: "whitesmoke", height: "100%" }}>Item 2</div>
    </Grid.Item>
    <Grid.Item>
      <div style={{ backgroundColor: "whitesmoke", height: "100%" }}>Item 3</div>
    </Grid.Item>
    <Grid.Item rowSpan={[1, 2, null, null, 1]}>
      <div style={{ backgroundColor: "whitesmoke", height: "100%" }}>Item 4</div>
    </Grid.Item>
    <Grid.Item>
      <div style={{ backgroundColor: "whitesmoke", height: "100%" }}>Item 5</div>
    </Grid.Item>
    <Grid.Item>
      <div style={{ backgroundColor: "whitesmoke", height: "100%" }}>Item 6</div>
    </Grid.Item>
    <Grid.Item>
      <div style={{ backgroundColor: "whitesmoke", height: "100%" }}>Item 7</div>
    </Grid.Item>
    <Grid.Item>
      <div style={{ backgroundColor: "whitesmoke", height: "100%" }}>Item 8</div>
    </Grid.Item>
    <Grid.Item>
      <div style={{ backgroundColor: "whitesmoke", height: "100%" }}>Item 9</div>
    </Grid.Item>
    <Grid.Item colSpan={[1, 2, 1]}>
      <div style={{ backgroundColor: "whitesmoke", height: "100%" }}>Item 10</div>
    </Grid.Item>
  </Grid>
);

WithCustomGridItems.parameters = {
  controls: { hideNoControlsWarning: true },
};
