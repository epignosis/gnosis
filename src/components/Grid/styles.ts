import { css, SerializedStyles } from "@emotion/react";
import { mq, breakpoints } from "@theme/utils/breakpoints";

const breakpointsKeys = Object.keys(breakpoints);

const getCSSProp = (prop: string, attr: string) => `${prop}: ${attr}`;

type GridProp = string | (number | null | undefined)[] | null;

const getCSSResponsiveProps = (prop: string, attrs: GridProp) => {
  if (typeof attrs === "string") return getCSSProp(prop, attrs);

  return attrs?.map((value, i) => {
    if (value) {
      return `
          ${mq[breakpointsKeys[i]]} {
            ${prop}: ${value};
          }
        `;
    }

    return "";
  });
};

const generateGridTemplates = (attrs: GridProp) => {
  const prop = "grid-template-columns";

  if (typeof attrs === "string") return `${prop}: repeat(${attrs}, 1fr)`;

  return attrs?.map((value, i) => {
    if (value) {
      return `
          ${mq[breakpointsKeys[i]]} {
            ${prop}: repeat(${value}, 1fr);
          }
        `;
    }

    return "";
  });
};

export const gridContainer = ({
  gap,
  templateColumns,
  columnGap,
  rowGap,
}: {
  gap: number;
  templateColumns: GridProp;
  columnGap: number | undefined;
  rowGap: number | undefined;
}): SerializedStyles => css`
  display: grid;
  ${gap && `grid-gap: ${gap}rem`};
  ${rowGap && `row-gap: ${rowGap}rem`};
  ${columnGap && `column-gap: ${columnGap}rem`};
  ${generateGridTemplates(templateColumns)}
`;

const generateGridSpan = (prop: string, attrs: GridProp) => {
  if (typeof attrs === "string") return `${prop}: span ${attrs}`;

  return attrs?.map((value, i) => {
    if (value) {
      return `
          ${mq[breakpointsKeys[i]]} {
            ${prop}: span ${value};
          }
        `;
    }

    return "";
  });
};

export const gridItem = ({
  colSpan,
  colStart,
  colEnd,
  rowSpan,
  rowStart,
  rowEnd,
}: {
  colSpan: GridProp;
  colStart: GridProp;
  colEnd: GridProp;
  rowSpan: GridProp;
  rowStart: GridProp;
  rowEnd: GridProp;
}): SerializedStyles => {
  return css`
    ${colSpan && generateGridSpan("grid-column", colSpan)};
    ${colStart && getCSSResponsiveProps("grid-column-start", colStart)};
    ${colEnd && getCSSResponsiveProps("grid-column-end", colEnd)};
    ${rowSpan && generateGridSpan("grid-row", rowSpan)};
    ${rowStart && getCSSResponsiveProps("grid-row-start", rowStart)};
    ${rowEnd && getCSSResponsiveProps("grid-row-end", rowEnd)};
  `;
};
