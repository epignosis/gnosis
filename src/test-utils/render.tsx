import React, { FC, ReactElement, ComponentType } from "react";
import { render, RenderResult, RenderOptions } from "@testing-library/react";
import ThemeProvider from "../components/ThemeProvider/ThemeProvider";

const AllTheProviders: FC = ({ children }) => <ThemeProvider>{children}</ThemeProvider>;

const customRender = (ui: ReactElement, options: RenderOptions): RenderResult =>
  render(ui, { wrapper: AllTheProviders as ComponentType, ...options });

export * from "@testing-library/react";

export { customRender as render };
