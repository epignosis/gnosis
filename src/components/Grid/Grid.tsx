import React, { FC } from "react";
import { gridContainer, gridItem } from "./styles";

type NumberOrNull = number | null;
type HTMLWrapperElement = "div" | "article" | "section";

export type ResponsiveValues =
  | [NumberOrNull, NumberOrNull, NumberOrNull?, NumberOrNull?, NumberOrNull?, NumberOrNull?]
  | number;

export type GridProps = {
  templateColumns: ResponsiveValues;
  gap?: number;
  columnGap?: number;
  rowGap?: number;
  as?: HTMLWrapperElement;
  className?: string;
};

type GridCompoundProps = {
  Item: FC<GridItemProps>;
};

const generateGridValues = (
  values: ResponsiveValues | "auto" | undefined,
): (number | null | undefined)[] | string | null => {
  switch (typeof values) {
    case "undefined":
      return null;
    case "string":
      return values;
    case "number":
      return values.toString();
    default:
      return values.map((value) => value);
  }
};

const Grid: FC<GridProps> & GridCompoundProps = (props) => {
  const { gap = 2, columnGap, rowGap, templateColumns, as = "div", className, children } = props;
  const templateColumnAttrs = generateGridValues(templateColumns);
  const Component = as;

  return (
    <Component
      css={gridContainer({ templateColumns: templateColumnAttrs, gap, columnGap, rowGap })}
      className={className}
    >
      {children}
    </Component>
  );
};

type GridItemProps = {
  colSpan?: ResponsiveValues | "auto";
  colStart?: ResponsiveValues | "auto";
  colEnd?: ResponsiveValues | "auto";
  rowSpan?: ResponsiveValues | "auto";
  rowStart?: ResponsiveValues | "auto";
  rowEnd?: ResponsiveValues | "auto";
  as?: HTMLWrapperElement;
  className?: string;
};

const Item: FC<GridItemProps> = (props) => {
  const {
    colSpan,
    colStart,
    colEnd,
    rowSpan,
    rowStart,
    rowEnd,
    as = "div",
    children,
    ...rest
  } = props;
  const colSpanAttrs = generateGridValues(colSpan);
  const colStartAttrs = generateGridValues(colStart);
  const colEndAttrs = generateGridValues(colEnd);
  const rowSpanAttrs = generateGridValues(rowSpan);
  const rowStartAttrs = generateGridValues(rowStart);
  const rowEndAttrs = generateGridValues(rowEnd);
  const Component = as;

  return (
    <Component
      css={gridItem({
        colSpan: colSpanAttrs,
        colStart: colStartAttrs,
        colEnd: colEndAttrs,
        rowSpan: rowSpanAttrs,
        rowStart: rowStartAttrs,
        rowEnd: rowEndAttrs,
      })}
      {...rest}
    >
      {children}
    </Component>
  );
};

Grid.Item = Item;

export default Grid;
