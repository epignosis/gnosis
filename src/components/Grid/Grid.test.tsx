import React from "react";
import { faker } from "@faker-js/faker";
import Grid from "./Grid";
import { render, screen } from "@test-utils/render";

describe("<Grid />", () => {
  it("renders correctly", () => {
    const paragraph = faker.lorem.paragraph();
    const { container } = render(
      <Grid templateColumns={5}>
        <Grid.Item>{paragraph}</Grid.Item>
        <Grid.Item>{paragraph}</Grid.Item>
        <Grid.Item>{paragraph}</Grid.Item>
        <Grid.Item>{paragraph}</Grid.Item>
        <Grid.Item>{paragraph}</Grid.Item>
      </Grid>,
    );

    const paragraphElements = screen.getAllByText(paragraph);

    expect(container).toBeInTheDocument();
    expect(container).toBeVisible();
    expect(paragraphElements).toHaveLength(5);
  });

  it("matches snapshot", () => {
    const { container } = render(
      <Grid id="grid-container-1" className="container" templateColumns={5}>
        <Grid.Item id="grid-item-1" className="grid-item">
          Test element
        </Grid.Item>
        <Grid.Item id="grid-item-2" className="grid-item">
          Test element
        </Grid.Item>
        <Grid.Item id="grid-item-3" className="grid-item">
          Test element
        </Grid.Item>
        <Grid.Item id="grid-item-4" className="grid-item">
          Test element
        </Grid.Item>
        <Grid.Item id="grid-item-5" className="grid-item">
          Test element
        </Grid.Item>
      </Grid>,
    );

    expect(container).toMatchSnapshot();
  });
});
