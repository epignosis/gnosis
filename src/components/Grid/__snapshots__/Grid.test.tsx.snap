// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Grid /> matches snapshot 1`] = `
<div>
  <div
    class="css-yqr4i5-gridContainer-gridContainer"
  >
    <div
      class="css-1is4t0f-gridItem-gridItem"
    >
      Test element
    </div>
    <div
      class="css-1is4t0f-gridItem-gridItem"
    >
      Test element
    </div>
    <div
      class="css-1is4t0f-gridItem-gridItem"
    >
      Test element
    </div>
    <div
      class="css-1is4t0f-gridItem-gridItem"
    >
      Test element
    </div>
    <div
      class="css-1is4t0f-gridItem-gridItem"
    >
      Test element
    </div>
  </div>
</div>
`;
